src/Sylius/Bundle/CoreBundle/EventListener/VariantOnhandListener.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+/*
+ * This file is part of the Shopblender package.
+ *
+ * (c) Shopblender
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\CoreBundle\EventListener;
+
+use SM\Factory\FactoryInterface;
+use Sylius\Component\Core\Model\OrderItemUnitInterface;
+use Sylius\Component\Core\Model\ProductVariantInterface;
+use Sylius\Component\Inventory\InventoryUnitTransitions;
+use Sylius\Component\Inventory\Repository\InventoryUnitRepositoryInterface;
+use Sylius\Component\Payment\Model\PaymentInterface;
+use Sylius\Component\Resource\Exception\UnexpectedTypeException;
+use Symfony\Component\EventDispatcher\GenericEvent;
+
+/**
+ * @author Robin Jansen <robinjansen51@gmail.com>
+ */
+class VariantOnhandListener
+{
+    /**
+     * @var OrderItemUnitRepository
+     */
+    protected $repository;
+
+    /**
+     * @var FactoryInterface
+     */
+    protected $factory;
+
+    /**
+     * VariantOnhandListener constructor.
+     *
+     * @param FactoryInterface $factory
+     * @param InventoryUnitRepositoryInterface $repository
+     */
+    public function __construct(FactoryInterface $factory, InventoryUnitRepositoryInterface $repository)
+    {
+        $this->repository = $repository;
+        $this->factory = $factory;
+    }
+
+    /**
+     * @param GenericEvent $event
+     */
+    public function fillBackorders(GenericEvent $event)
+    {
+        $variant = $this->getProductVariant($event);
+        $onHand = $variant->getOnHand();
+
+        if ($onHand <= 0) {
+            return;
+        }
+
+        /** @var OrderItemUnitInterface[] $units */
+        $units = $this->repository->findByStockableAndInventoryState(
+            $variant,
+            OrderItemUnitInterface::STATE_BACKORDERED,
+            $onHand
+        );
+
+        foreach ($units as $unit) {
+            $paymentState = $unit->getOrderItem()->getOrder()->getPaymentState();
+            $sm = $this->factory->get($unit, InventoryUnitTransitions::GRAPH);
+
+            if ($paymentState === PaymentInterface::STATE_COMPLETED) {
+                $sm->apply(InventoryUnitTransitions::SYLIUS_SELL);
+            } else {
```

Comment:
Can we get rid off this nasty `else`? Maybe some simple private method and early return?